packages/modules/psa/api.py

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python3
+
+from typing import Union
+import logging
+import os
+import json
+import urllib
+import requests
+import base64
+from modules.common.store import RAMDISK_PATH
+# currently unused
+# import sys
+# import time
+# import getopt
+# import getpass
+# import http.client as http_client
+
+log = logging.getLogger("soc."+__name__)
+
+# these two lines enable debugging at httplib level (requests->urllib3->http.client)
+# You will see the REQUEST, including HEADERS and DATA, and RESPONSE with HEADERS but without DATA.
+# The only thing missing will be the response.body which is not logged.
+
+
+def dump_json(data: dict, fout: str):
+    replyFile = str(RAMDISK_PATH) + fout
+    try:
+        f = open(replyFile, 'w', encoding='utf-8')
+    except Exception as e:
+        log.debug("psa.dump_json: chmod File" + replyFile + ", exception, e=" + str(e))
+        os.system("sudo rm " + replyFile)
+        f = open(replyFile, 'w', encoding='utf-8')
+    json.dump(data, f, ensure_ascii=False, indent=4)
+    f.close()
+
+
+def requestPSAData(userid: str,
+                   password: str,
+                   client_id: str,
+                   client_secret: str,
+                   manufacturer: str,
+                   soccalc: str,
+                   vehicle: int) -> dict:
+
+    # Step1: Authentication: get access_token
+    userpass = client_id + ':' + client_secret
+    encoded_u = base64.b64encode(userpass.encode()).decode()
+    # header1 = str('Basic %s' % encoded_u)
+    scope = 'openid profile'
+    if (manufacturer == "Peugeot"):
+        brand = 'peugeot.com'
+        realm = 'clientsB2CPeugeot'
+    elif (manufacturer == "Citroen"):
+        brand = 'citroen.com'
+        realm = 'clientsB2CCitroen'
+    elif (manufacturer == "DS"):
+        brand = 'driveds.com'
+        realm = 'clientsB2CDS'
+    elif (manufacturer == "Opel"):
+        brand = 'opel.com'
+        realm = 'clientsB2COpel'
+    elif (manufacturer == "Vauxhall"):
+        brand = 'vauxhall.co.uk'
+        realm = 'clientsB2CVauxhall'
+    # vin = '?'
+    data = {'realm': realm, 'grant_type': 'password', 'password': password, 'username': userid, 'scope': scope}
+    headers = {'Content-Type': 'application/x-www-form-urlencoded', 'Authorization': 'Basic %s' % encoded_u}
+    reg = 'https://idpcvs.' + brand + '/am/oauth2/access_token'
+    dmp = {}   # dump request 1
+    dmp['reg'] = reg
+    dmp['data'] = data
+    dmp['headers'] = headers
+    dump_json(dmp, "/psareq1vh" + str(vehicle))
+    response = requests.post(reg, data=data, headers=headers)
+    responsetext = response.text
+    responsestatus = response.status_code
+    dmp = {}   # dump response 1
+    dmp['text'] = str(responsetext.encode("utf-8"))
+    dmp['status'] = str(responsestatus)
+    dump_json(dmp, "/psareply1vh" + str(vehicle))
+    psa_config = json.loads(responsetext)
```

Comment:
```suggestion
    data = {'realm': realm, 'grant_type': 'password', 'password': password, 'username': userid, 'scope': scope}
    headers = {'Content-Type': 'application/x-www-form-urlencoded', 'Authorization': 'Basic %s' % encoded_u}  
    req.get_http_session().post(f'https://idpcvs.{brand}/am/oauth2/access_token', data=data, headers=headers).json()
```
Siehe auch mein Kommentar zu PR #638. Ich habe gesehen, dass Du nicht der ursprüngliche Autor des Codes bist. Nichtsdestotrotz tut dem Code ein Refactoring vor der Portierung aus 1.9 an dieser Stelle gut.

Reply:
Die weiteren Änderungen müsste mrinas machen.
Ich werde ihn bitten, sich das mal anzusehen.

Reply:
Ja, das kann ich gern übernehmen. @LKuemmel können wir den PR erstmal so übernehmen wie der ist? Sobald der PR gemerged ist kann ich die Anpassungen vornehmen. Alternativ könnte ich die neuen Dateien hier rauskopieren und einen neuen PR einstellen wenn euch das lieber ist.

Reply:
Bitte einen neuen PR erstellen.

Reply:
okay kein Problem, ist in Arbeit.